Catch exceptions in main

SE-FDr opened the pull request.

SE-FDr committed:
catch exceptions in main, ensure a sane exit

SE-FDr committed:
typo

luepges commented on monticore-generator/src/main/resources/_cli/Main.ftl:
```diff
@@ -1,5 +1,17 @@
+              + " during the execution of the ${grammarname}Tool."
+              + System.lineSeparator() + "This error is unexpected"
+              + " and does not indicate an issue with any provided models.",
+          exception
```
Das Ganze führt jetzt zu dem Unterschied, dass Errors in der Ausführung mit Exit Code 0 (aka "alles gut") beendet werden.
Wenn wir den Change schon machen -> Rückgabewert zu int Ändern und fail=1/okay=0 zurückgeben?

SE-FDr replied:
Wie möchtest du den Rückgabewert von `main` ändern?
Ich würde `System.exit` eher vermeiden wollen, es sei denn, wir entfernen (lese: passen an) jeden Test, der zurzeit direkt die `main` Methode testet.

SE-FDr replied:
Discussed in MC Weekly:
Add `System.exit`, change all the tests that call `main` to call `run` instead.

@luepges our gradle plugins should be fine?

SE-FDr commented on monticore-generator/src/main/resources/_cli/Main.ftl:
```diff
@@ -1,5 +1,17 @@
+              + System.lineSeparator() + "This error is unexpected"
+              + " and does not indicate an issue with any provided models.",
+          exception
+      );
```
```suggestion
      );
      System.exit(1);
```

SE-FDr committed:
better exit code on exceptions

SE-FDr commented:
requires https://github.com/MontiCore/monticore/pull/236

luepges commented on monticore-test/01.experiments/automaton/src/test/java/AutomataToolTest.java:
```diff
@@ -76,10 +77,15 @@ public void executeHierarchyPingPong() {
   }
 
+  // cannot be tested, as main has been changed to call System.exit.
+  @Disabled
```
Auch nicht in nem Mock env?

luepges replied:
In hindsight: Not really

luepges commented on monticore-generator/src/main/resources/_cli/Main.ftl:
```diff
@@ -1,5 +1,18 @@
-   tool.run(args);
+    try {
+      ${grammarname}Tool tool = new  ${grammarname}Tool();
+      tool.run(args);
```
Fehlt hier am Ende nicht noch ein System.exit(1) if Log.hasErrors?

SE-FDr committed:
Exit with 1 if errors occurred. Ensure exit.

SE-FDr committed:
simplify main()

luepges commented on monticore-generator/src/main/resources/_cli/Main.ftl:
```diff
@@ -1,5 +1,20 @@
+    }
+    // properly exit with a code
+    Log.ensureInitialization();
+    System.exit(Log.getErrorCount() == 0 ? 0 : 1);
```
This might break a bunch of tests using Tool#main instead of "new Tool(...)".

I'm actually not against this breaking change, as it makes it clear that the main method entry is only meant for CLI tools.

SE-FDr replied:
This was my intention.
`run` should be the entry point (iff one wants the same behavior).

This is also the reason why I replaced the calls to `XTool::main` with `run` in our main projects.

SE-FDr committed:
document Tool ::main and ::run

SE-FDr committed:
test that main does not output exceptions

# Conflicts:
#	monticore-test/01.experiments/automaton/src/test/java/AutomataToolTest.java

SE-FDr committed:
test that main does not output exceptions

# Conflicts:
#	monticore-test/01.experiments/automaton/src/test/java/AutomataToolTest.java

luepges reviewed (approved):
(wenn das
